File: 19-RemoveNthNodeFromEndofList/removeNthNodeFromEndofList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode temp = new ListNode(1);
        temp.next = head;
        ListNode near = temp;
        ListNode far = temp;
        //define near and far with distance n+1. In other words, when far -> null (after end), near -> (n+1)-th node from end.
        for (int i = 0; i< n+1; i++){
            far = far.next;
        }
        while(far != null) {
            near = near.next;
            far = far.next;
        }
        near.next = near.next.next;
        return temp.next;
}
}

File: 19-RemoveNthNodeFromEndofList/removeNthNodeFromEndofList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {

        ListNode* p2=head;
        ListNode* dummy=new ListNode(0);
        dummy->next=head;
        ListNode* p1=dummy;
        
        for(int i=0;i<n;i++)    p2=p2->next;
        while(p2){
            p1=p1->next;
            p2=p2->next;
        }
        p1->next=p1->next->next;
        return dummy->next;
    }
};